pkg/action/build-uki.go

```diff
@@ -0,0 +1,538 @@
+		"dd",
+		"mkfs.msdos",
+		"mmd",
+		"mcopy",
+		"xorriso",
+	}
+
+	neededFiles := []string{
+		// TODO: this should come from the given image, not the OS where enki runs
+		"/usr/lib/systemd/boot/efi/linuxx64.efi.stub",
+	}
+
+	for _, b := range neededBinaries {
+		_, err := exec.LookPath(b)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, b := range neededFiles {
+		_, err := os.Stat(b)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *BuildUKIAction) setupDirectoriesAndFiles(tmpDir string) error {
+	if err := os.Symlink("/usr/bin/immucore", filepath.Join(tmpDir, "init")); err != nil {
+		return fmt.Errorf("error creating symlink: %w", err)
+	}
+
+	// able to mount oem under here if found
+	if err := os.MkdirAll(filepath.Join(tmpDir, "oem"), os.ModeDir); err != nil {
+		return fmt.Errorf("error creating /oem dir: %w", err)
+	}
+
+	// mount the esp under here if found
+	if err := os.MkdirAll(filepath.Join(tmpDir, "efi"), os.ModeDir); err != nil {
+		return fmt.Errorf("error creating /oem dir: %w", err)
+	}
+
+	// for install/upgrade they copy stuff there
+	if err := os.MkdirAll(filepath.Join(tmpDir, "usr/local/cloud-config"), os.ModeDir); err != nil {
+		return fmt.Errorf("error creating /oem dir: %w", err)
+	}
+
+	return nil
+}
+
+// createInitramfs creates a compressed initramfs file (cpio format, gzipped).
+// The resulting file is named "initrd" and is saved inthe sourceDir.
+func (b *BuildUKIAction) createInitramfs(sourceDir string) error {
+	format := "newc"
+	archiver, err := cpio.Format(format)
+	if err != nil {
+		return fmt.Errorf("format %q not supported: %w", format, err)
+	}
+
+	cpioFileName := filepath.Join(sourceDir, "initramfs.cpio")
+	cpioFile, err := os.Create(cpioFileName)
+	if err != nil {
+		return fmt.Errorf("creating cpio file: %w", err)
+	}
+	defer cpioFile.Close()
+
+	rw := archiver.Writer(cpioFile)
+	cr := cpio.NewRecorder()
+
+	// List of directories to exclude
+	excludeDirs := map[string]bool{
+		"./sys":  true,
+		"./run":  true,
+		"./dev":  true,
+		"./tmp":  true,
+		"./proc": true,
+	}
+
+	if err = os.Chdir(sourceDir); err != nil {
+		return fmt.Errorf("changing to %s directory: %w", sourceDir, err)
+	}
+
+	// Walk through the source directory and add files to the cpio archive
+	err = filepath.Walk(".", func(filePath string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Check if the current directory should be excluded
+		if excludeDirs[filePath] {
+			return filepath.SkipDir
+		}
+
+		rec, err := cr.GetRecord(filePath)
+		if err != nil {
+			return fmt.Errorf("getting record of %q failed: %w", filePath, err)
+		}
+
+		rec.Name = strings.TrimPrefix(rec.Name, sourceDir)
+		if err := rw.WriteRecord(rec); err != nil {
+			return fmt.Errorf("writing record %q failed: %w", filePath, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error walking the source dir: %w", err)
+	}
+
+	if err := cpio.WriteTrailer(rw); err != nil {
```

Comment:
Commands in this method were mostly copied from here: https://github.com/u-root/u-root/blob/25faeb4566c826ea11bc725f291af8f46d893cc5/cmds/core/cpio/cpio.go#L133-L145